pkg/keyspace/tso_keyspace_group.go

```diff
@@ -72,42 +122,137 @@ func (m *GroupManager) GetKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup,
 		err error
 	)
 
-	m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+	if err := m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
 		kg, err = m.store.LoadKeyspaceGroup(txn, id)
 		if err != nil {
 			return err
 		}
 		return nil
-	})
+	}); err != nil {
+		return nil, err
+	}
 	return kg, nil
 }
 
 // DeleteKeyspaceGroupByID deletes the keyspace group by id.
 func (m *GroupManager) DeleteKeyspaceGroupByID(id uint32) error {
-	m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+	var (
+		kg  *endpoint.KeyspaceGroup
+		err error
+	)
+
+	if err := m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+		kg, err = m.store.LoadKeyspaceGroup(txn, id)
+		if err != nil {
+			return err
+		}
 		return m.store.DeleteKeyspaceGroup(txn, id)
-	})
+	}); err != nil {
+		return err
+	}
+
+	userKind := endpoint.StringUserKind(kg.UserKind)
+	m.Lock()
+	defer m.Unlock()
+	// TODO: move out the keyspace to another group
+	// we don't need the keyspace group as the return value
+	m.groups[userKind].Remove(id)
+
 	return nil
 }
 
-func (m *GroupManager) saveKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup) error {
+func (m *GroupManager) saveKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup, overwrite bool) error {
 	return m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
 		for _, keyspaceGroup := range keyspaceGroups {
```

Comment:
do we need read lock for keyspaceGroups here?

Reply:
use lock now